### PR TITLE
[AD-175713] Change utility to timer trigger

### PR DESCRIPTION
**Pull request by JHUDSON-SMALL-DFE**
ADO: https://sfa-gov-uk.visualstudio.com/National%20Careers%20Service%20(NCS)/_workitems/edit/175713

### CODE REVIEW

File: NCS.DSS.GDPRUpdateCustomerInformation/NCS.DSS.GDPRUtility.csproj [hunk 1]
```diff
@@ -7,6 +7,8 @@
   </PropertyGroup>
   <ItemGroup>
     <PackageReference Include="Azure.Identity" Version="1.13.1" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.23.0" />
```
Comment (bmatthews-DFE) [inline NCS.DSS.GDPRUpdateCustomerInformation/NCS.DSS.GDPRUtility.csproj:10]: Uninstall this package - it's not used

File: NCS.DSS.GDPRUpdateCustomerInformation.Tests/NCS.DSS.GDPRUtility.Tests.csproj [hunk 2]
```diff
@@ -15,6 +15,8 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
     <PackageReference Include="FakeItEasy" Version="8.3.0" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.23.0" />
```
Comment (bmatthews-DFE) [inline NCS.DSS.GDPRUpdateCustomerInformation.Tests/NCS.DSS.GDPRUtility.Tests.csproj:18]: Uninstall this package - it's not used

File: NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs [hunk 3]
```diff
@@ -1,26 +1,26 @@
...
+        public GDPRUtility(IIdentifyAndAnonymiseDataService identifyAndAnonymiseDataService, ILogger<GDPRUtility> logger)
         {
             _identifyAndAnonymiseDataService = identifyAndAnonymiseDataService;
             _logger = logger;
         }
 
-        [Function(nameof(GDPRUpdateCustomerInformation))]
-        public async Task<IActionResult> RunAsync([HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = null)] HttpRequest req)
+        [Function(nameof(GDPRUtility))]
+        [FixedDelayRetry(3, "00:00:15")]
```
Comment (bmatthews-DFE) [inline NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs:20]: Please add a comment to describe what this does

File: NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs [hunk 4]
```diff
@@ -1,26 +1,26 @@
...
+        public async Task<IActionResult> RunAsync([TimerTrigger("0 0 2 1 * *")] TimerInfo timer)
```
Comment (bmatthews-DFE) [inline NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs:21]: Please add a comment to explain what this trigger means (NCRONTAB)
`{second} {minute} {hour} {day} {month} {day-of-week}`

**Review by bmatthews-DFE — CHANGES_REQUESTED** [review]
See comments

File: NCS.DSS.GDPRUpdateCustomerInformation.Tests/NCS.DSS.GDPRUtility.Tests.csproj (hunk 2 above)
Comment (JHUDSON-SMALL-DFE) [reply NCS.DSS.GDPRUpdateCustomerInformation.Tests/NCS.DSS.GDPRUtility.Tests.csproj:18]: it's required for the timer

File: NCS.DSS.GDPRUpdateCustomerInformation/NCS.DSS.GDPRUtility.csproj (hunk 1 above)
Comment (bmatthews-DFE) [reply NCS.DSS.GDPRUpdateCustomerInformation/NCS.DSS.GDPRUtility.csproj:10]: Pretty sure `Microsoft.Azure.Functions.Worker.Extensions.Timer` provides the timer functionality, not `Microsoft.Azure.Functions.Worker` - try to uninstall it and rebuilt - it should still succeed

File: NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs (hunk 4 above)
Comment (bmatthews-DFE) [reply NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs:27]: Resolving - thanks Joe

File: NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs (hunk 3 above)
Comment (bmatthews-DFE) [reply NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs]: Resolving - thanks Joe

**Review by bmatthews-DFE — CHANGES_REQUESTED** [review]
Further comments

File: NCS.DSS.GDPRUpdateCustomerInformation.Tests/NCS.DSS.GDPRUtility.Tests.csproj (hunk 2 above)
Comment (JHUDSON-SMALL-DFE) [reply NCS.DSS.GDPRUpdateCustomerInformation.Tests/NCS.DSS.GDPRUtility.Tests.csproj:18]: as above, it's required for the isolated worker model: https://learn.microsoft.com/en-us/azure/azure-functions/dotnet-isolated-process-guide

File: NCS.DSS.GDPRUpdateCustomerInformation/NCS.DSS.GDPRUtility.csproj (hunk 1 above)
Comment (bmatthews-DFE) [reply NCS.DSS.GDPRUpdateCustomerInformation/NCS.DSS.GDPRUtility.csproj:10]: It's not required: 
image

This was deployed to AT without that package - please uninstall

File: NCS.DSS.GDPRUpdateCustomerInformation.Tests/NCS.DSS.GDPRUtility.Tests.csproj (hunk 2 above)
Comment (bmatthews-DFE) [reply NCS.DSS.GDPRUpdateCustomerInformation.Tests/NCS.DSS.GDPRUtility.Tests.csproj:18]: It's not required: 
image

This was deployed to AT without that package - please uninstall

File: NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs [hunk 5]
```diff
@@ -1,26 +1,32 @@
...
-        [Function(nameof(GDPRUpdateCustomerInformation))]
-        public async Task<IActionResult> RunAsync([HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = null)] HttpRequest req)
+        /*
+        In case of failed execution, FixedDelayRetry will retry up to three times after a 15 second delay interval.
+        GDPRUtility runs at 2am every month (on the 1st of the month) as defined in NCRONTAB syntax:
+        {seconds} {minutes} {hours} {day of month} {month} {day-of-week}
+        */
+
+        [Function(nameof(GDPRUtility))]
+        [FixedDelayRetry(3, "00:00:15")]
```
Comment (bmatthews-DFE) [inline NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs:26]: Do we want to retry it on failure?

File: NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs [hunk 6]
```diff
@@ -1,26 +1,32 @@
...
+        public GDPRUtility(IIdentifyAndAnonymiseDataService identifyAndAnonymiseDataService, ILogger<GDPRUtility> logger)
         {
             _identifyAndAnonymiseDataService = identifyAndAnonymiseDataService;
             _logger = logger;
         }
 
-        [Function(nameof(GDPRUpdateCustomerInformation))]
-        public async Task<IActionResult> RunAsync([HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = null)] HttpRequest req)
+        /*
+        In case of failed execution, FixedDelayRetry will retry up to three times after a 15 second delay interval.
```
Comment (bmatthews-DFE) [inline NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs:20]: Thanks for adding - can you please add a link to this page: https://crontab.cronhub.io/ as it is a NCRONTAB calculator

Comment (bmatthews-DFE) [reply NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs:20]: image

**Review by bmatthews-DFE — CHANGES_REQUESTED** [review]
See comments

File: NCS.DSS.GDPRUpdateCustomerInformation.Tests/NCS.DSS.GDPRUtility.Tests.csproj (hunk 2 above)
Comment (JHUDSON-SMALL-DFE) [reply NCS.DSS.GDPRUpdateCustomerInformation.Tests/NCS.DSS.GDPRUtility.Tests.csproj]: It's building for me today... how strange that it says it was required

File: NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs (hunk 5 above)
Comment (JHUDSON-SMALL-DFE) [reply NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs:26]: Generally I would say so for something automated of this nature - may reduce the amount of investigations required in future if we can see that a 2nd attempt worked as expected.

**Review by bmatthews-DFE — CHANGES_REQUESTED** [review]
Final outstanding comment to resolve

File: NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs (hunk 5 above)
Comment (JHUDSON-SMALL-DFE) [reply NCS.DSS.GDPRUpdateCustomerInformation/Function/GDPRUtility.cs]: sure

**Review by bmatthews-DFE — DISMISSED** [review]
LGTM ✅

**Review by hnandra — APPROVED** [review]

**Pull request merged**